openapi-codegen/core/src/main/scala/sttp/tapir/codegen/ClassDefinitionGenerator.scala

```diff
@@ -34,40 +62,27 @@ class ClassDefinitionGenerator {
-        s"""implicit def seqCodec[T: com.github.plokhotnyuk.jsoniter_scala.core.JsonValueCodec]: com.github.plokhotnyuk.jsoniter_scala.core.JsonValueCodec[List[T]] =
-        |  com.github.plokhotnyuk.jsoniter_scala.macros.JsonCodecMaker.make[List[T]]($jsoniterDefaultConfig)
-        |implicit def optionCodec[T: com.github.plokhotnyuk.jsoniter_scala.core.JsonValueCodec]: com.github.plokhotnyuk.jsoniter_scala.core.JsonValueCodec[Option[T]] =
-        |  com.github.plokhotnyuk.jsoniter_scala.macros.JsonCodecMaker.make[Option[T]]($jsoniterDefaultConfig)
-        |""".stripMargin
-      else ""
+    val adtTypes = adtInheritanceMap.flatMap(_._2).toSeq.distinct.map(name => s"sealed trait $name").mkString("", "\n", "\n")
     val enumQuerySerdeHelper = if (!generatesQueryParamEnums) "" else enumQuerySerdeHelperDefn(targetScala3)
-    // For jsoniter-scala, we define explicit serdes for any 'primitive' params (e.g. List[java.util.UUID]) that we reference.
-    // This should be the set of all json param refs not included in our schema definitions
-    val additionalExplicitSerdes = jsonParamRefs.toSeq
-      .filter(x => !allSchemas.contains(x))
-      .map(s =>
-        jsonSerdeLib match {
-          case JsonSerdeLib.Jsoniter =>
-            val name = s.replace("[", "_").replace("]", "_").replace(".", "_") + "JsonCodec"
-            s"""implicit lazy val $name: com.github.plokhotnyuk.jsoniter_scala.core.JsonValueCodec[$s] =
-            |  com.github.plokhotnyuk.jsoniter_scala.macros.JsonCodecMaker.make[$s]($jsoniterDefaultConfig)""".stripMargin
-          case _ => ""
-        }
-      )
-      .mkString("", "\n", "\n")
+    val postDefns = JsonSerdeGenerator.serdeDefs(doc, jsonSerdeLib, jsonParamRefs, allTransitiveJsonParamRefs, fullModelPath)
     val defns = doc.components
       .map(_.schemas.flatMap {
         case (name, obj: OpenapiSchemaObject) =>
-          generateClass(allSchemas, name, obj, jsonSerdeLib, allTransitiveJsonParamRefs)
+          generateClass(allSchemas, name, obj, allTransitiveJsonParamRefs, adtInheritanceMap)
         case (name, obj: OpenapiSchemaEnum) =>
           generateEnum(name, obj, targetScala3, queryParamRefs, jsonSerdeLib, allTransitiveJsonParamRefs)
-        case (name, OpenapiSchemaMap(valueSchema, _)) => generateMap(name, valueSchema, jsonSerdeLib, allTransitiveJsonParamRefs)
+        case (name, OpenapiSchemaMap(valueSchema, _)) => generateMap(name, valueSchema)
+        case (_, _: OpenapiSchemaOneOf)               => Nil
         case (n, x) => throw new NotImplementedError(s"Only objects, enums and maps supported! (for $n found ${x})")
       })
       .map(_.mkString("\n"))
-    defns.map(additionalExplicitSerdes + maybeJsonSerdeHelpers + enumQuerySerdeHelper + _)
+    val helpers = (enumQuerySerdeHelper + adtTypes).linesIterator
+      .filterNot(_.forall(_.isWhitespace))
+      .mkString("\n")
+    // Jsoniter-scala ADT serdes need to live in a separate file from the class defns
```

Comment:
Yeah, fair enough. I think it'd need a tiny bit of fiddling because of the possibility for endpoints to go into the 'base' object but in general I kinda prefer splitting the serdes into a separate file too

Reply:
So I hadn't realised that it was legal, but it turns out that even with the discriminated jsoniter-scala ADT serdes, you can get away with importing the serdes into the endpoints and the endpoints into the serdes mutually-recursively. So I've split the json serdes out into a separate file under all circumstances, and just import it back into the base file if necessary (i.e. if any untagged endpoints are defined).